adapters/dmx/dmx.go

```diff
@@ -0,0 +1,333 @@
+	var errs []error
+
+	if http.StatusNoContent == response.StatusCode {
+		return nil, nil
+	}
+
+	if http.StatusBadRequest == response.StatusCode {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code 400"),
+		}}
+	}
+
+	if http.StatusOK != response.StatusCode {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected response no status code"),
+		}}
+	}
+
+	var bidResp openrtb.BidResponse
+
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(5)
+
+	for _, sb := range bidResp.SeatBid {
+		for i := range sb.Bid {
+			bidType, err := getMediaTypeForImp(sb.Bid[i].ImpID, request.Imp)
+			if err != nil {
+				errs = append(errs, err)
+			} else {
+				b := &adapters.TypedBid{
+					Bid:     &sb.Bid[i],
+					BidType: bidType,
+				}
+				if b.BidType == openrtb_ext.BidTypeVideo {
+					b.Bid.AdM = videoImpInsertion(b.Bid)
+				}
+				bidResponse.Bids = append(bidResponse.Bids, b)
+			}
+		}
+	}
+	return bidResponse, errs
+
+}
+
+func fetchParams(params dmxExt, inst openrtb.Imp, ins openrtb.Imp, imps []openrtb.Imp, banner *openrtb.Banner, video *openrtb.Video, intVal int8) []openrtb.Imp {
+	if params.Bidder.TagId != "" {
+		ins = openrtb.Imp{
+			ID:     inst.ID,
+			TagID:  params.Bidder.TagId,
+			Ext:    inst.Ext,
+			Secure: &intVal,
+		}
+	}
+
+	if params.Bidder.DmxId != "" {
+		ins = openrtb.Imp{
+			ID:     inst.ID,
+			TagID:  params.Bidder.DmxId,
+			Ext:    inst.Ext,
+			Secure: &intVal,
+		}
+	}
+	if banner != nil {
+		ins.Banner = banner
+	}
+
+	if video != nil {
+		ins.Video = video
+	}
+
+	if ins.TagID == "" {
+		return imps
+	}
+	imps = append(imps, ins)
+	return imps
+}
+
+func addParams(str string) string {
+	if str != "" {
+		return "?sellerid=" + url.QueryEscape(str)
+	}
+	return ""
+}
+
+func getMediaTypeForImp(impID string, imps []openrtb.Imp) (openrtb_ext.BidType, error) {
+	mediaType := openrtb_ext.BidTypeBanner
+	for _, imp := range imps {
+		if imp.ID == impID {
+			if imp.Banner == nil && imp.Video != nil {
+				mediaType = openrtb_ext.BidTypeVideo
+			}
+			return mediaType, nil
+		}
+	}
+
+	// This shouldnt happen. Lets handle it just incase by returning an error.
+	return "", &errortypes.BadInput{
+		Message: fmt.Sprintf("Failed to find impression \"%s\" ", impID),
+	}
+}
+
+func videoImpInsertion(bid *openrtb.Bid) string {
+	adm := bid.AdM
+	nurl := bid.NURL
+	search := "</Impression>"
+	imp := "</Impression><Impression><![CDATA[%s]]></Impression>"
+	wrapped_nurl := fmt.Sprintf(imp, nurl)
+	results := strings.Replace(adm, search, wrapped_nurl, 1)
+	return results
+}
+
+func isDmxParams(t interface{}) bool {
+	switch t.(type) {
+	case dmxParams:
+		return true
+	default:
+		return false
+	}
+}
+
+func CloneValue(source interface{}, destin interface{}) {
```

Comment:
I'm not exactly sure what this code is doing. It looks like it makes a shallow copy since it doesn't appear to be looping over the members (unless the reflect packages magically handle that?). The request object passed in to this method is already a shallow copy, so this likely isn't necessary. The code you added to copy App and Site structures before mutating them solves the shared memory problem in the recommended manner.

Reply:
Yes, this does seem odd to have since you use it only once, and you know the type of the source/destination when you use it. What does this give you over a simple:
`reqcopy := *request`
`dmxReq = &reqcopy`
And do you need `dmxReq` to be a pointer rather than an actual struct?